_episodes/16-parallel.md

```diff
@@ -90,21 +433,149 @@ clearly. Press "q" to exit the pager program (`less`).
-memory footprint of 11.2 GB: if your machine has less RAM than that, it will grind
-to a halt. If you have 16 GB installed, you won't quite make it to 1&frac12; billion points.
+```
+comm = MPI.COMM_WORLD
+cpus = comm.Get_size()
+rank = comm.Get_rank()
+```
+{: .language-python}
+
+immediately before the `n_samples` line to set up the MPI environment for
+each process.
+
+We replace the `start_time` and `counts` lines with the lines:
+
+```
+if rank == 0:
+  start_time = datetime.datetime.now()
+  partitions = [ int(n_samples / cpus) ] * cpus
+  counts = [ int(0) ] * cpus
+else:
+  partitions = None
+  counts = None
+```
+{: .language-python}
+
+to ensure that only the rank 0 process measures times and coordinates
+the work to be distributed to all the ranks, and that the other ranks
+get placeholder values for the `partitions` and `counts` variables.
+
+Immediately below these lines, we'll add the following three lines:
+
+```
+partition_item = comm.scatter(partitions, root=0)
+count_item = inside_circle(partition_item)
+counts = comm.gather(count_item, root=0)
+```
+{: .language-python}
+
+to:
+* distribute the work among the ranks with MPI `scatter`,
+* call the `inside_circle` function so each rank can perform its share
+  of the work,
+* collect each rank's results into a `counts` variable on rank 0 using MPI
+  `gather`.
+
+Illustrations of these steps are shown below.
+
+---
+
+Setup the MPI environment and initialize local variables &mdash; including the
+vector containing the number of points to generate on each parallel processor:
+
+{% include figure.html url="" max-width="25%" file="/fig/initialize.png" 
+   alt="MPI initialize" caption="" %}
+
+Distribute the number of points from the originating vector to all the parallel
+processors:
+
+{% include figure.html url="" max-width="25%" file="/fig/scatter.png" 
+   alt="MPI scatter" caption="" %}
+
+Perform the computation in parallel:
+
+{% include figure.html url="" max-width="25%" file="/fig/compute.png"
+   alt="MPI compute" caption="" %}
+
+Retrieve counts from all the parallel processes:
+
+{% include figure.html url="" max-width="28%" file="/fig/gather.png"
+   alt="MPI gather" caption="" %}
+
+Print out the report:
+
+{% include figure.html url="" max-width="30%" file="/fig/finalize.png"
+   alt="MPI finalize" caption="" %}
+
+---
+
+Finally, we'll ensure the `my_pi` through `print` lines only run on rank
+0 by placing them inside an `if` statement as:
+
+```
+if rank == 0:
+   my_pi = 4.0 * sum(counts) / sum(partitions)
+   end_time = datetime.datetime.now()
+   elapsed_time = (end_time - start_time).total_seconds()
+   size_of_float = np.dtype(np.float64).itemsize
+   memory_required = 3 * sum(partitions) * size_of_float / (1024**3)
+   print("Pi: {}, memory: {} GiB, time: {} s".format(my_pi, memory_required, elapsed_time))
+```
+{: .language-python}
+
+The resulting MPI Python script is:
```

Comment:
Once more, let's not print the whole thing. Perhaps links for the instructor to the completed script, in case someone has fallen hopelessly behind, but we want live-coding rather than copy-paste.